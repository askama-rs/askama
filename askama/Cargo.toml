[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
