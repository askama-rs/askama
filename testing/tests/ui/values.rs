use askama::Template;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = "a"|value %}{% endif %}"#,
)]
struct A;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = "a"|value::<u8, u8> %}{% endif %}"#,
)]
struct B;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = askama::get_value("a") %}{% endif %}"#,
)]
struct C;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = askama::get_value::<u8, u8>("a") %}{% endif %}"#,
)]
struct D;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = askama::get_value::<u8>() %}{% endif %}"#,
)]
struct E;

#[derive(Template)]
#[template(
    ext = "html",
    source = r#"{% if let Ok(x) = askama::get_value::<u8>("a", "b") %}{% endif %}"#,
)]
struct F;

fn main() {}
