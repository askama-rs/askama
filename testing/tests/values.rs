use std::any::Any;
use std::collections::HashMap;

use askama::Template;

#[test]
fn test_values() {
    #[derive(Template)]
    #[template(
        source = r#"{% if let Ok(bla) = "a" | value::<u32> %}{{bla}}{% endif %}"#,
        ext = "txt"
    )]
    struct V;

    let mut values: HashMap<String, Box<dyn Any>> = HashMap::default();
    assert_eq!(V.render_with_values(&values).unwrap(), "");
    values.insert("a".to_string(), Box::new(12u32));
    assert_eq!(V.render_with_values(&values).unwrap(), "12");
    values.insert("a".to_string(), Box::new(false));
    assert_eq!(V.render_with_values(&values).unwrap(), "");
}

#[test]
fn test_values2() {
    #[derive(Template)]
    #[template(
        source = r#"
            {%- if let Ok(bla) = "a" | value::<&str> %}{{bla}}{% endif -%}
            {%- if let Ok(bla) = "b" | value::<bool> %} {{bla}}{% endif -%}
        "#,
        ext = "txt"
    )]
    struct V;

    let mut values: HashMap<String, Box<dyn Any>> = HashMap::default();
    assert_eq!(V.render_with_values(&values).unwrap(), "");
    values.insert("a".to_string(), Box::new("hey"));
    assert_eq!(V.render_with_values(&values).unwrap(), "hey");
    values.insert("b".to_string(), Box::new(false));
    assert_eq!(V.render_with_values(&values).unwrap(), "hey false");
    values.remove("a");
    assert_eq!(V.render_with_values(&values).unwrap(), " false");
}

#[test]
fn test_values3() {
    #[derive(Template)]
    #[template(
        source = r#"
            {%- match "data" | value::<&str> -%}
                {%- when Ok(data) -%} ok={{ data }}
                {%- when Err(err) -%} err={{ err }}
            {%- endmatch -%}
        "#,
        ext = "txt"
    )]
    struct V;

    let mut values: HashMap<String, Box<dyn Any>> = HashMap::default();
    assert_eq!(
        V.render_with_values(&values).unwrap(),
        "err=key missing in values"
    );
    values.insert("data".to_string(), Box::new(false));
    assert_eq!(
        V.render_with_values(&values).unwrap(),
        "err=value has wrong type"
    );
    values.insert("data".to_string(), Box::new("hey"));
    assert_eq!(V.render_with_values(&values).unwrap(), "ok=hey");
    values.insert("data".to_string(), Box::new(Box::new("hey")));
    assert_eq!(V.render_with_values(&values).unwrap(), "ok=hey");
}

#[test]
fn test_value_function_getter() {
    #[derive(Template)]
    #[template(
        source = r#"{% if let Ok(bla) = askama::get_value::<u32>("a") %}{{bla}}{% endif %}"#,
        ext = "txt"
    )]
    struct V;

    let mut values: HashMap<String, Box<dyn Any>> = HashMap::default();
    assert_eq!(V.render_with_values(&values).unwrap(), "");
    values.insert("a".to_string(), Box::new(12u32));
    assert_eq!(V.render_with_values(&values).unwrap(), "12");
    values.insert("a".to_string(), Box::new(false));
    assert_eq!(V.render_with_values(&values).unwrap(), "");
}
