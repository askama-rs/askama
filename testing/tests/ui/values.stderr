error: `value` filter expects one generic, found 0
 --> A.html:1:18
       "\"a\"|value %}{% endif %}"
 --> tests/ui/values.rs:6:14
  |
6 |     source = r#"{% if let Ok(x) = "a"|value %}{% endif %}"#,
  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `value` filter expects one generic, found 2
 --> B.html:1:18
       "\"a\"|value::<u8, u8> %}{% endif %}"
  --> tests/ui/values.rs:13:14
   |
13 |     source = r#"{% if let Ok(x) = "a"|value::<u8, u8> %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `get_value` function expects one generic, found 0
 --> C.html:1:18
       "askama::get_value(\"a\") %}{% endif %}"
  --> tests/ui/values.rs:20:14
   |
20 |     source = r#"{% if let Ok(x) = askama::get_value("a") %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `get_value` function expects one generic, found 2
 --> D.html:1:18
       "askama::get_value::<u8, u8>(\"a\") %}{% endif %}"
  --> tests/ui/values.rs:27:14
   |
27 |     source = r#"{% if let Ok(x) = askama::get_value::<u8, u8>("a") %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `get_value` function only takes one argument, found 0
 --> E.html:1:18
       "askama::get_value::<u8>() %}{% endif %}"
  --> tests/ui/values.rs:34:14
   |
34 |     source = r#"{% if let Ok(x) = askama::get_value::<u8>() %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `get_value` function only takes one argument, found 2
 --> F.html:1:18
       "askama::get_value::<u8>(\"a\", \"b\") %}{% endif %}"
  --> tests/ui/values.rs:41:14
   |
41 |     source = r#"{% if let Ok(x) = askama::get_value::<u8>("a", "b") %}{% endif %}"#,
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
